### PR TITLE
 Add isAppender functionality on mobile

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description

This PR is adding `isAppender` functionality to mobile `MediaPlaceholder` component.

Related gutenberg-mobile [PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1327)

image

See [wordpress-mobile/gutenberg-mobile#1299](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1299)

According to the design, `isAppender` button should be displayed below the gallery once an item in the gallery is selected, that's why information, whether a block is selected, has to be passed from parent component into `media-placeholder` component.

Web implementation was not touched.

## How has this been tested?

Tested manually in local gutenberg using `gallery`, however, currently it's hard to test it, so I have to do some tweaks in `image` component to treat it as `gallery` since the gallery component is not prepared for mobile yet.

Generally, my goal was to adjust `image` component and do some code changes within `image/ edit.native.js` to behave similarly to `gallery` component. To accomplish that I've changed:

* added `isAppender` prop and make it conditional on `hasImage` value
* added `isSelected` prop
* if there is not image return `mediaPlaceholder` like in gallery

```
		const hasImage = !! url;

		const mediaPlaceholder = (
				<MediaPlaceholder
					allowedTypes={ [MEDIA_TYPE_IMAGE] }
					onSelect={ this.onSelectMediaUploadOption }
					icon={ this.getIcon( false ) }
					onFocus={ this.props.onFocus }
					isAppender={ hasImage }
					disableMediaButtons={ !isSelected }
				/>
		)

		if ( ! hasImage ) {
			return mediaPlaceholder;
		}
```

then `image` component is returning a chosen photo along with `mediaPlacholder` the same as in `gallery`

```
		return (
			<MediaUpload
				allowedTypes={ [MEDIA_TYPE_IMAGE ]}
				onSelect={ this.onSelectMediaUploadOption }
				render={ ( { open, getMediaOptions } ) => {
					return (
						<>
							{getImageComponent( open, getMediaOptions )}
							{mediaPlaceholder}
						</>
					)
				} }
			/>
		);
```

## Types of changes

* Support for `isAppender` prop on mobile platform
* New prop `isSelected` for mobile to distinguish if block is selected

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by pinarol** [comment]
Could you please update the target branch as `rnmobile/master` on this PR as well? We have a blocker [issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1332) that prevents us moving fwd with the master branch.

File: packages/block-editor/src/components/media-placeholder/README.md
Comment (pinarol) [inline packages/block-editor/src/components/media-placeholder/README.md]: It doesn't look like `isSelected` needs to be part of MediaPlaceholder API. Block can decide to render/not render the MediaPlaceholder due to its selected state. 
Also as far as I see web also handles it this way without needing to add this to MediaPlaceholder API and we want to come close to web as much as possible.
Let me know if I am missing something.

Comment (lukewalczak) [reply packages/block-editor/src/components/media-placeholder/README.md]: According to the mobile gallery [design](https://app.zeplin.io/project/5bfecbea71ce881122eadd2f/screen/5d3b1273384a876ce6c195b3), `isAppender` button should be displayed only when `gallery` block is selected, that's why there is a need to pass `isSelected` value from parent component into `mediaPlaceholder`. Please, correct me if I understood it wrongly.

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/README.md]: Could you be able to do some investigation about how this works on web? As far as I see mediaplaceholder isn't rendered if block is not selected on web as well. So how is it handled without some isSelected prop? We want to keep props of components as similar as possible between platforms. 
Independent from this, it doesn't sound right for a component to have a prop named 'isSelected' where it actually decides rendering/not rendering it.

Comment (lukewalczak) [reply packages/block-editor/src/components/media-placeholder/README.md]: Sure, will investigate it!

Comment (lukewalczak) [reply packages/block-editor/src/components/media-placeholder/README.md]: After the investigation, I've noticed that web media-placeholder accepts prop called `dropZoneUIOnly={ hasImages && ! isSelected }` which then within placeholder is rendering gallery with appender when `dropZoneUIOnly === false` otherwise without appender.

Comment (lukewalczak) [reply packages/block-editor/src/components/media-placeholder/README.md]: Maybe instead of adding new prop `isSelected` I should reuse and document prop `dropZoneUIOnly`, wdyt?

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/README.md]: maybe we can make use of the same prop for parity, although, we don't currently have drag & drop of media. But it looks like a nice enhancement for the future. wdyt @koke ? I think we can start using `dropZoneUIOnly` on mobile too.

Comment (koke) [reply packages/block-editor/src/components/media-placeholder/README.md]: I don't think `dropZoneUIOnly` is the right prop to use here. I checked the `README` and it's not documented anywhere. From what I'm seeing, as long as media upload is possible, the drop zone would be rendered on the web, and the UI to pick media is the variable here.

Adding `dropZoneUIOnly` to mobile would imply that we support drag and drop, and could be confusing. I'd rather flip the logic and have a `showMediaSelectionUI` (or another better name) prop that is set only when the block is selected.

Comment (lukewalczak) [reply packages/block-editor/src/components/media-placeholder/README.md]: The name you've proposed (`showMediaSelectionUI`) sounds good for me. Just to be sure, would you like to replace `dropZoneUIOnly` with `showMediaSelectionUI` on the web and have the same prop on mobile?

Comment (koke) [reply packages/block-editor/src/components/media-placeholder/README.md]: Yes, I would propose that change for the web as well

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/README.md]: LGTM as soon as we keep the parity. 👍

File: packages/block-editor/src/components/media-placeholder/index.js
Comment (pinarol) [inline packages/block-editor/src/components/media-placeholder/index.js]: @gziolo 👋 would changing the prop name here be a problem? Are there some steps we need to take before changing it?

Comment (gziolo) [reply packages/block-editor/src/components/media-placeholder/index.js]: It'd be a breaking change as this prop is part of the public API.

The corresponding PR was merged in March (#12367) and WordPress 5.2 went out on May 7th (https://wordpress.org/news/2019/05/jaco/). I don't see it in `wp/5.2` branch so it means it was merged during the beta phase and not included in the release cycle.

Anyway, it's been a few months since it was exposed as part of the public API so we would have to keep the old name as well to ensure we don't break any plugins which might be using it.

Comment (gziolo) [reply packages/block-editor/src/components/media-placeholder/index.js]: I see it documented in the `master` branch but not here:

https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/components/media-placeholder#dropzoneuionly

> If true, only the Drop Zone will be rendered. No UI controls to upload the media will be shown. The `disableDropZone` prop still takes precedence over `dropZoneUIOnly` – specifying both as true will result in nothing to be rendered.

@jorgefilipecosta - can you help here?

Comment (lukewalczak) [reply packages/block-editor/src/components/media-placeholder/index.js]: So, as far as I understand we are not able to change that prop name. 
Should we then share the `dropZoneUIOnly ` prop (which sounds weird a bit on mobile) between platforms?

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: When I look at the latest [docs](https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/components/media-placeholder#dropzoneuionly) I see that disableDropZone and dropZoneUIOnly are also related: `The disableDropZone prop still takes precedence over dropZoneUIOnly – specifying both as true will result in nothing to be rendered` Thus, just renaming it as `showMediaSelectionUI` won't make sense unless we refactor it to work independently.
AFAIU, we'll either need to change the behavior of web implementation too or we'll keep the props same but we'll need to add to docs that drag & drop is not yet supported on mobile. I am aiming towards to the latter one since that's a nice feature to have on mobile too. wdyt @koke ?

Comment (koke) [reply packages/block-editor/src/components/media-placeholder/index.js]: Those look a bit confusing to me, and I'd prefer to flip them both 😅 

- `showMediaSelectionUI` (`== ! dropZoneUIOnly`)
- `showDropZone` (`== ! disableDropZone`)

It's still possible to set them both as `false` and have it render nothing, but I think it's clearer that that would happen from these names, and the props wouldn't seem to contradict each other.

> So, as far as I understand we are not able to change that prop name

What I understand is that we still need to support the old props. We can leave only the new ones in the documentation and all the uses across the codebase, but the implementation should still look for the old props and behave accordingly, in case a plugin is using those.

Comment (koke) [reply packages/block-editor/src/components/media-placeholder/index.js]: > that's a nice feature to have on mobile too(in the future)

Definitely agree. While it's not yet on the roadmap, we should support drag and drop media upload on mobile at some point

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: As far as I understand, disabling the drop zone is a very rare case. So I find the `disableDropZone` style configuration more suitable in this case, because if we convert it to `showDropZone` we'll either need to:
- set this to true in almost all places that uses this component 
or 
- give a default true inside the component which wouldn't be as intuitive as the current usage

On the other hand I agree with the `showMediaSelectionUI` change but this time I'd expect this value wouldn't be overridden by `disableDropZone`. Because the new name conveys that it should be responsible for just showing/hiding the media selection UI, it is not conveying any info about the relation with the drop zone. I understand that making these 2 independent is just the thing we want but I really don't know how we can do that transition in a smooth way. In the deprecation phase we'll need to support both new and old ones together, but `showMediaSelectionUI` will be acting differently so it brings a certain complexity to the logic.

So if we are going to add drag&drop in the following 3 months for example, does it worth the effort? That's the thing I am trying to figure. I mean it is up to us to plan for this.

Comment (koke) [reply packages/block-editor/src/components/media-placeholder/index.js]: > give a default true inside the component which wouldn't be as intuitive as the current usage

I don't understand why that wouldn't be intuitive. I'm fine with leaving the name, since we don't really need to change it for this, as all we need is what you said here:

> I understand that making these 2 independent is just the thing we want

I think we're fine on that front? If we have `showMediaSelectionUI` and `disableDropZone`, those would be independent:

showMediaSelectionUI | disableDropZone | renders
---------------------|-----------------|--------
false                | false           | drop zone
false                | true            | nothing
true                 | false           | media selection
true                 | true            | drop zone + media selection

I think this is the same behavior as before, but previously the name `dropZoneUIOnly` would seem to contradict `disableDropZone` when both were true.

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: I thought currently we'd see nothing if we set:

````
disableDropZone={ true } 
dropZoneUIOnly={ false }
```

because I was under the impression that disableDropZone will disable all the things related to dropZone since it has precedence. But it looks like it is not the case, I did a quick experiment in my local:

<img width="916" alt="Screen Shot 2019-09-05 at 19 57 14" src="https://user-images.githubusercontent.com/5032900/64363254-cedea100-d018-11e9-8e6a-be1a6ae39bbd.png">

So basically they do act independently in this sense, just like how we wanted. This makes transition phase simple because we won't need to deal with changing the logic. What is more, I think new name makes more sense with the current behavior imho. But ofc we need to get this reviewed.

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: So let's try to list some steps to propose such a change properly for a review:

1. Add showMediaSelectionUI
2. Deprecate `dropZoneUIOnly`
3. Use these together for a while, like:

before:

``` 
 if ( dropZoneUIOnly )
``` 

after:

```
 if ( dropZoneUIOnly &&  ! showMediaSelectionUI )
``` 

Anything else I am missing?

BTW I've never deprecated a prop from components before but I guess it looks like [this](https://github.com/WordPress/gutenberg/blob/a491a79ab2ec9c05b296c50e1ce0c5e389e19175/packages/block-editor/src/components/rich-text/index.js#L297)

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: Another thing worth noting might be dropZoneUIOnly is representing almost the MediaPlaceholder itself, I mean if I set it to true I see nothing:

Screen Shot 2019-09-05 at 20 55 29

So again I am not sure about using a positive keyword that has a default `true` value and will be true in almost all the cases. It feels more natural to me to use sth like `hidePlaceholder` with a default `false`. Because when I see showXXX I question myself like do I need to set this to true explicitly and I feel the need to check the docs. On the other hand with a name like `hidePlaceholder` or `hideMediaSelectionUI` I am sure at first sight that it won't get hidden until I set it to `true`. That was the reason I said it was more intuitive. Not a big deal but just wanted to mention.

Comment (koke) [reply packages/block-editor/src/components/media-placeholder/index.js]: That makes sense. For consistency with the other prop I would suggest disableMediaSelection instead of “hide”

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: `disableMediaSelection` sounds good to me 👍

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: @jorgefilipecosta let us know if you have concerns about this. Otherwise I think we can go ahead and deprecate `dropZoneUIOnly` and add `disableMediaSelection`?

Comment (lukewalczak) [reply packages/block-editor/src/components/media-placeholder/index.js]: @pinarol 	I've updated, my PR according to the conversation above:

* added `disableMediaSelection` prop
* deprecate`dropZoneUIOnly` prop

File: packages/block-editor/src/components/media-placeholder/README.md
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/media-placeholder/README.md]: We have a button to select from the media library, a button to upload a new media file and sometimes a button to insert from a URL. I guess given that the property name is disableMediaSelection users may think it only disables the media selection button while leaving the upload button.
I'm not sure of what a good name would be, maybe `disableMediaUI`, `disableMediaButtons`.

File: packages/block-editor/src/components/media-placeholder/index.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/media-placeholder/index.js]: I think we have a bug here, even if the new prop disableMediaSelection is used we always call the deprecated function. It should be called if dropZoneUIOnly is used but not called if disableMediaSelection is used.

File: packages/e2e-tests/specs/demo.test.js
Comment (jorgefilipecosta) [inline packages/e2e-tests/specs/demo.test.js]: We should not expect the console to warn, I think if we address the deprecated call problem, we can remove this expects and we will also fix the end 2 end test.

File: packages/block-editor/src/components/media-placeholder/README.md
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/media-placeholder/README.md]: Should we specify the default value (false) as part of the docs?

File: packages/block-editor/src/components/media-placeholder/README.md
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/media-placeholder/README.md]: I think this flag if enabled will render a drop zone but not a placeholder we can update the text to: "Whether render a dropzone without any other additional UI".

**Comment by jorgefilipecosta** [comment]
Hi @lukewalczak, @pinarol, @koke I reviewed this PR, and I think the props update is an improvement and makes things more consistent 👍. I'm not sure if the new prop name is ideal more details in https://github.com/WordPress/gutenberg/pull/17195#discussion_r326230755 I left two alternatives, but I'm open to other suggestions.

**Comment by pinarol** [comment]
Thanks for the review @jorgefilipecosta ! Both of your suggestions look good to me. I don’t have a strong preferance but disableMediaButtons feels slightly more clear to me.

**Comment by lukewalczak** [comment]
Hello @pinarol @jorgefilipecosta, I've refactored a code along with documentation according to your suggestion (thanks!).

File: packages/block-editor/src/components/media-placeholder/index.native.js
Comment (pinarol) [inline packages/block-editor/src/components/media-placeholder/index.native.js]: it looks like there's sth wrong here, it is returning null if disableMediaButtons is undefined or false.

**Review by pinarol — COMMENTED** [review]
It looks like the PR is broken after last changes, let's make sure we test both mobile and web side after every change.

failing scenario on mobile:
if I set disableMediaButtons={ true } I can see the appender
if I set disableMediaButtons={ false } I can’t see it
or if it is not defined at all, again I can’t see it

that should've been the opposite

File: packages/block-editor/src/components/media-placeholder/index.js
Comment (pinarol) [inline packages/block-editor/src/components/media-placeholder/index.js]: do we need the additional `if ( dropZoneUIOnly || disableMediaButtons )` ?

Comment (pinarol) [reply packages/block-editor/src/components/media-placeholder/index.js]: ignore, I got the order of curl braces wrongly

**Comment by lukewalczak** [comment]
@pinarol You're right - after changing the prop name from the first approach (`isSelected`) to the current one (`disableMediaButtons`) condition should be opposite. Thanks!

**Review by pinarol — APPROVED** [review]
Looks good to me! 🎉 

Tested both with iOS example app and local gutenberg web instance. Although I think we should wait for @jorgefilipecosta 's approval to merge.

**Review by jorgefilipecosta — APPROVED** [review]
Tested the web version and it works well, the PR seems ready.
There is only one step we should do before the merge, add a "Breaking Change" changelog entry in packages/block-editor/CHANGELOG.md specifying the dropZoneUIOnly prop was deprecated and disableMediaButtons should now be used instead.

File: packages/block-editor/CHANGELOG.md
Comment (gziolo) [inline packages/block-editor/CHANGELOG.md]: How about:

```suggestion
- `dropZoneUIOnly` prop in `MediaPlaceholder` components has been deprecated in favor of `disableMediaButtons` prop.
```

**Comment by gziolo** [comment]
> add a "Breaking Change" changelog entry

@jorgefilipecosta It's only a deprecation for the time being which triggers minor version updated. Only when functionality is completely removed, it's considered a breaking change. Regardless, thank you for ensuring we have an entry included in the changelog :)

**Comment by gziolo** [comment]
@lukewalczak, great work and congrats on your first code contribution to Gutenberg 🎉

**Pull request merged**
